### ISSUE TITLE
`assert_never` raises `reportArgumentType` when case should be unreachable.

### ISSUE THREAD

**Issue by wangkev** [microsoft/pyright#7435]
**Describe the bug**
The following snippet

```python
from typing import assert_never

def foo(a: int | float, b: bool | str) -> str:
    if isinstance(a, int) and not isinstance(b, bool):
        raise TypeError('a is int and b is not bool')
    if isinstance(a, float) and not isinstance(b, str):
        raise TypeError('a is float and b is not str')
    
    match a, b:
        case int(), bool():
            return("int and bool")
        case float(), str():
            return("float and str")
        case _ as unreachable:
            assert_never(unreachable)
```
raises the following error:
```
Argument of type "tuple[int | float, bool | str]" cannot be assigned to parameter "__arg" of type "Never" in function "assert_never"
  Type "tuple[int | float, bool | str]" cannot be assigned to type "Never"  (reportArgumentType)
```
I would expect `assert_never` to be unreachable because either `a` is `int` and `b` is `bool` or `a` is `float` and `b` is `str` from the first two checks.

**VS Code extension or command-line**
Running in pyright playground.

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

The `isinstance` checks at the top of this function do nothing to narrow the type of either `a` or `b`, so their types remain `int | float` and `bool | str` respectively. Type narrowing is performed on individual variables, not on collections of variables.

The subject of the match is therefore of type `tuple[int | float, bool | str]`. Because tuples are covariant and immutable, they can be "expanded" as described in [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/tuples.html#tuple-type-form). However, the number of subtypes in an expanded tuple grows combinatorially. For this reason, pyright limits tuple expansion to cases where only one element of the tuple is expanded. For example, if you change the code to the following, it will type check without error:

```python
def foo(a: int | float, b: bool | str) -> str:
    match a, b:
        case int(), x:
            assert isinstance(x, bool), "a is int and b is not bool"
            return "int and bool"
        case float(), y:
            assert isinstance(y, str), "a is float and b is not str"
            return "float and str"
        case _ as unreachable:
            assert_never(unreachable)
```
